src/library_glemu.js

```diff
@@ -2108,6 +2133,10 @@ var LibraryGLEmulation = {
         enabledAttributesKey = (enabledAttributesKey << 1) | (GLEmulation.lightingEnabled ? GLEmulation.lightEnabled[lightId] : 0);
       }
 
+      // By alpha testing mode
+      enabledAttributesKey = (enabledAttributesKey << 1) | GLEmulation.alphaTestEnabled;
+      enabledAttributesKey = (enabledAttributesKey << 3) | (GLEmulation.alphaTestFunc - 0x200);
```

Comment:
Do we have enough bits here, I wonder? Reading the above existing code, it looks like the number of bits we need depends on the number of lights and clip planes, so maybe this was always potentially close to the limit. But this adds a requirement for 4 more bits, which may break existing code, unless I'm missing something? I don't have a sense for how likely that is in practice, though - ?

Reply:
I spent some time thinking about this and trying some tricky ways to squeeze all state bits into the 31 bits available. But all that added a lot of unnecessary code complexity. So instead I added an additional level to the maptree and split the enabledAttributesKey into two.
On my machine the MAX_TEXTURES was getting set to 32 which on itself would be too many state bits. Since no code relying on Immediate mode should ever use that many textures I limited it to 28 to make it always fit into 31 bits (3 bits for pos, color, normal attributes).